qualtran/drawing/graphviz.py

```diff
@@ -399,3 +400,31 @@ def cxn_edge(self, left_id: str, right_id: str, cxn: Connection) -> pydot.Edge:
             arrowhead='dot',
             arrowsize=0.25,
         )
+
+
+class TypedGraphDrawer(PrettyGraphDrawer):
+    @staticmethod
+    def _fmt_dtype(dtype: QDType):
+        label = f'{dtype.__class__.__name__}({dtype.num_qubits})'
+        return label
+
+    def cxn_label(self, cxn: Connection) -> str:
+        """Overridable method to return labels for connections."""
+
+        l, r = cxn.left.reg.dtype, cxn.right.reg.dtype
+        if l == r:
+            return self._fmt_dtype(l)
+        else:
+            return f'{self._fmt_dtype(l)}-{self._fmt_dtype(r)}'
```

Comment:
We will have a cast bloq but this specific example is because the splits and joins were not updated to use the appropriate type (#722). In the future this could be relaxed for the case when consistent implicit casts enforced, but it could be helpful for visualizing where we are being loose and strict with types.